source/extensions/filters/http/cache/cache_filter.cc

```diff
@@ -76,65 +143,77 @@ Http::FilterHeadersStatus CacheFilter::encodeHeaders(Http::ResponseHeaderMap& he
 }
 
 Http::FilterDataStatus CacheFilter::encodeData(Buffer::Instance& data, bool end_stream) {
+  if (skip_encoding_) {
+    return Http::FilterDataStatus::Continue;
+  }
   if (insert_) {
-    ENVOY_STREAM_LOG(debug, "CacheFilter::encodeHeaders inserting body", *encoder_callbacks_);
+    ENVOY_STREAM_LOG(debug, "CacheFilter::encodeData inserting body", *encoder_callbacks_);
     // TODO(toddmgreer): Wait for the cache if necessary.
     insert_->insertBody(
         data, [](bool) {}, end_stream);
+  } else if (encode_cached_response_state_ == EncodeCachedResponseState::IterationStopped) {
+    // Iteration stopped waiting for cached body (and trailers) to be encoded
+    return Http::FilterDataStatus::StopIterationAndBuffer;
   }
   return Http::FilterDataStatus::Continue;
 }
 
+void CacheFilter::getHeaders() {
+  ASSERT(lookup_, "CacheFilter is trying to call getHeaders with no LookupContext");
+  lookup_->getHeaders([this](LookupResult&& result) { onHeaders(std::move(result)); });
+}
+
+void CacheFilter::getBody() {
+  ASSERT(lookup_, "CacheFilter is trying to call getBody with no LookupContext");
+  ASSERT(!remaining_body_.empty(), "No reason to call getBody when there's no body to get.");
+  lookup_->getBody(remaining_body_[0],
+                   [this](Buffer::InstancePtr&& body) { onBody(std::move(body)); });
+}
+
+void CacheFilter::getTrailers() {
+  ASSERT(lookup_, "CacheFilter is trying to call getTrailers with no LookupContext");
+  ASSERT(response_has_trailers_, "No reason to call getTrailers when there's no trailers to get.");
+  lookup_->getTrailers(
+      [this](Http::ResponseTrailerMapPtr&& trailers) { onTrailers(std::move(trailers)); });
+}
+
 void CacheFilter::onHeaders(LookupResult&& result) {
   // TODO(yosrym93): Handle request only-if-cached directive
   switch (result.cache_entry_status_) {
   case CacheEntryStatus::FoundNotModified:
   case CacheEntryStatus::UnsatisfiableRange:
     NOT_IMPLEMENTED_GCOVR_EXCL_LINE; // We don't yet return or support these codes.
   case CacheEntryStatus::RequiresValidation:
-    // Cache entries that require validation are treated as unusable entries
-    // until validation is implemented
-    // TODO(yosrym93): Implement response validation
+    // If a cache entry requires validation inject validation headers in the request and let it
+    // pass through as if no cache entry was found. If the cache entry was valid the response
+    // status should be 304 (unmodified), and the cache entry will be injected in the response body
+    validating_cache_entry_ = true;
+    lookup_result_ = std::make_unique<LookupResult>(std::move(result));
+    injectValidationHeaders();
+    // The following line is necessary for GCC to know that there is no break on purpose
```

Comment:
There is ABSL_FALLTHROUGH_INTENDED, although... I only see uses here in Envoy in a couple extensions?

Reply:
After searching the codebase I found that they have their own `FALLTHRU` macro at [source/common/common/macros.h](https://github.com/envoyproxy/envoy/blob/master/source/common/common/macros.h). Will switch to using it.